src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,539 @@
+        {
+            var response = AwsHttpClientHelper.GetEC2ResponseAsync().GetAwaiter().GetResult();
+            var parsedResponse = BsonDocument.Parse(response);
+            var username = parsedResponse.GetValue("AccessKeyId", null)?.AsString;
+            var password = parsedResponse.GetValue("SecretAccessKey", null)?.AsString;
+            var sessionToken = parsedResponse.GetValue("Token", null)?.AsString;
+
+            var awsCredentials = new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+
+            ValidateAwsCredentials(awsCredentials);
+
+            if (awsCredentials.Username == null)
+            {
+                return null;
+            }
+
+            return awsCredentials;
+        }
+
+        private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            if (properties == null)
+            {
+                return null;
+            }
+
+            foreach (var pair in properties)
+            {
+                if (pair.Key.ToUpperInvariant() == "AWS_SESSION_TOKEN")
+                {
+                    return pair.Value;
+                }
+            }
+
+            return null;
+        }
+
+        private static SecureString ToSecureString(string str)
+        {
+            if (str == null)
+            {
+                return null;
+            }
+
+            var secureString = new SecureString();
+            foreach (var c in str)
+            {
+                secureString.AppendChar(c);
+            }
+            secureString.MakeReadOnly();
+
+            return secureString;
+        }
+
+        private static void ValidateAwsCredentials(AwsCredentials awsCredentials)
+        {
+            if (awsCredentials.Username == null && (awsCredentials.Password != null || awsCredentials.SessionToken != null))
+            {
+                throw new InvalidOperationException("A MONGODB-AWS credential must have an access key ID.");
+            }
+            if (awsCredentials.Username != null && awsCredentials.Password == null)
+            {
+                throw new InvalidOperationException("A MONGODB-AWS credential must have a secret access key.");
+            }
+        }
+
+        private static void ValidateMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            if (properties == null)
+            {
+                return;
+            }
+
+            foreach (var pair in properties)
+            {
+                if (pair.Key.ToUpperInvariant() != "AWS_SESSION_TOKEN")
+                {
+                    throw new ArgumentException($"Unknown AWS property '{pair.Key}'.", nameof(properties));
+                }
+            }
+        }
+
+        // nested classes
+        private class AwsCredentials
```

Comment:
Make this class a little safer by making it immutable and adding some validation in the constructor:

    private class AwsCredentials
    {
        private readonly string _accessKeyId;
        private readonly SecureString _secretAccessKey;
        private readonly string _sessionToken;

        public AwsCredentials(string accessKeyId, SecureString secretAccessKey, string sessionToken)
        {
            _accessKeyId = Ensure.IsNotNull(accessKeyId, nameof(accessKeyId));
            _secretAccessKey = Ensure.IsNotNull(secretAccessKey, nameof(secretAccessKey));
            _sessionToken = sessionToken; // can be null
        }

        public string AccessKeyId => _accessKeyId;
        public SecureString SecretAccessKey => _secretAccessKey;
        public string SessionToken => _sessionToken;
    }